internal/controllers/machinedeployment/machinedeployment_status.go

```diff
@@ -0,0 +1,502 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machinedeployment
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/internal/controllers/machinedeployment/mdutil"
+	"sigs.k8s.io/cluster-api/util/collections"
+	v1beta2conditions "sigs.k8s.io/cluster-api/util/conditions/v1beta2"
+	clog "sigs.k8s.io/cluster-api/util/log"
+)
+
+func (r *Reconciler) updateStatus(ctx context.Context, s *scope) (retErr error) {
+	// Get all Machines controlled by this MachineDeployment.
+	var machines, machinesToBeRemediated, unHealthyMachines collections.Machines
+	if selectorMap, err := metav1.LabelSelectorAsMap(&s.machineDeployment.Spec.Selector); err == nil {
+		machineList := &clusterv1.MachineList{}
+		if err := r.Client.List(ctx, machineList, client.InNamespace(s.machineDeployment.Namespace), client.MatchingLabels(selectorMap)); err != nil {
+			retErr = errors.Wrap(err, "failed to list machines")
+		}
+		machines = collections.FromMachineList(machineList)
+		machinesToBeRemediated = machines.Filter(collections.IsUnhealthyAndOwnerRemediated)
+		unHealthyMachines = machines.Filter(collections.IsUnhealthy)
```

Comment:
Let's make this an else branch. I have no idea what guarantees CR gives for the content of MachineList when List returns an error

Reply:
Also I think without the else branch we wouldn't hit the `machine == nil` branch in setMachinesReadyCondition.

But about that part. Can we please set an additional bool when we hit the error case here and then use that in setMachinesReadyCondition, setMachinesUpToDateCondition, setRemediatingCondition ... instead of relying on a difference between `machines == nil` & `len(machines) ==0`